### PR TITLE
Make AppVeyor run Coverity scan

### LINKED ISSUE #986: Leverage Coverity scan on AppVeyor
Related https://github.com/appveyor/ci/issues/144

### PR DESCRIPTION
**Pull request by nulltoken**
Fix #986

### CODE REVIEW

**Comment by whoisj** [comment]
:tada: :+1:

**Comment by nulltoken** [comment]
@whoisj It's not there yet... :wink:

**Comment by whoisj** [comment]
oh I was commenting on the concept - not actually approving anything :confused:

**Comment by nulltoken** [comment]
@FeodorFitsner I think I'd need some help from you with this PR :pray:

How can publish the file `libgit2sharp.zip` as an artifact?

038e9623c0f2a9934333c4035077c2fec8bd8742 doesn't publish it.
Neither 5459aefb25d27f305b7a0a3c85cc99b54bdb1191

Although it exists in the folder (see the `ls` output in the build log)

I'm surely missing something obvious here, your insight would be very much appreciated.

**Comment by FeodorFitsner** [comment]
When is `libgit2sharp.zip` created? Artifacts are getting packaged after `test` and before `deploy` stages. Maybe you should use script to push artifacts: http://www.appveyor.com/docs/packaging-artifacts#pushing-artifacts-from-scripts

File: appveyor.yml [hunk 1]
```diff
@@ -55,22 +59,58 @@ before_build:
...
+
+      & nuget install PublishCoverity -Version 0.9.0 -ExcludeVersion -OutputDirectory .\packages
+
+      # Write-Host -ForegroundColor "Green" "Building project with Coverity..."
+      # & cov-build.exe --dir cov-int "C:\Program Files (x86)\MSBuild\12.0\bin\msbuild.exe" $env:APPVEYOR_BUILD_FOLDER\LibGit2Sharp.sln /p:Configuration=Release /logger:"C:\Program Files\AppVeyor\BuildAgent\Appveyor.MSBuildLogger.dll"
+      Write-Host -ForegroundColor "Green" "Compressing Coverity results..."
+
+      & .\packages\PublishCoverity\PublishCoverity.exe compress `
+        -i "$env:APPVEYOR_BUILD_FOLDER\cov-int" `
+        -o "$env:APPVEYOR_BUILD_FOLDER\$env:APPVEYOR_PROJECT_NAME.zip" `
```
Comment (nulltoken) [inline appveyor.yml]: @FeodorFitsner The file is created by this call to `PublishCoverity.exe`

Comment (FeodorFitsner) [inline appveyor.yml]: OK, it's `on_success` stage what means it's too late for picking up artifacts. Use `Push-AppVeyorArtifact` cmdlet: http://www.appveyor.com/docs/build-worker-api#push-artifact

**Comment by nulltoken** [comment]
>Artifacts are getting packaged after `test` and before `deploy` stages.

@FeodorFitsner Perfect! They're currently created in the `on_success` stage. I'm going to move this to `build_success`, then.

As always, you're spot on with an amazing support. :heart: :bangbang:

**Comment by nulltoken** [comment]
@FeodorFitsner s/build_success/deploy_script/

**Comment by nulltoken** [comment]
@FeodorFitsner I eventually relied on `after_test` (BTW, maybe the doc should mention that this `after_*` hooks are invoked if the related stage succeeded.

**Comment by nulltoken** [comment]
@csMACnz `PublishCoverity` is awesome!

However, Coverity may reject a build because the maximum number of uploaded builds has been exceeded. How could we nicely wrap the call to `PublishCoverity.exe publish` so that it doesn't fail the AppVeyor build when that happens?

For instance, this current builds fails because the AppVeyor secured variables aren't valued in a PR. According to you, how should we tweak the script so that the build doesn't fail any more?

/cc @FeodorFitsner

**Comment by whoisj** [comment]
Can we limit Covertity to only run after all other unit tests pass?

**Comment by nulltoken** [comment]
@whoisj The tests are currently commented out to speed the build. Of course, they'll be reinstated once I'm done and this PR is ready for review.

**Comment by nulltoken** [comment]
@whoisj I took an alternate route.

AppVeyor exposes a way to **[schedule builds](http://www.appveyor.com/docs/build-configuration#scheduled-builds)**.

New proposal is to run Coverity once a week during a scheduled build. As the build will be a re-run of the the latest vNext tip, there's no need to run the tests any longer (this saves us from failed builds because of network hiccups).

Thoughts?

**Comment by whoisj** [comment]
That works so long as somebody is responsive to the Covertity report and creates issues appropriately; and release points are at, or as close as reasonable to, Covertity scan points.

**Comment by nulltoken** [comment]
> That works so long as somebody is responsive to the Coverity report and creates issues appropriately; and release points are at, or as close as reasonable to, Covertity scan points.

Indeed, the idea would be to fix them as quickly as possible.
FWIW, the way it's configured, found defects are not disclosed by default.

**Comment by nulltoken** [comment]
So, good people, good to merge?

**Comment by whoisj** [comment]
:+1:

**Pull request merged**